### PR TITLE
Modularize and extract mergeq

### PR DESCRIPTION
**Pull request by dbalatero**
### mergeq

- [x] don't commit TCH credentials
- [x] remove guard
- [x] remove zeus
- [x] move repo-specific mergeq data to a `.mergeq` directory
- [x] move `.merging` to `.mergeq/merging`
- [x] get rid of `run_hooks`
- [x] make per-repo named hooks for specific hook points
- [x] test queuing a merge into this repo's `staging` branch
- [x] go back to the branch you were working on after push
- [x] document the hooks
- [x] write setup instructions in the README 

### mergeq_ci

- [x] test a full cycle

### CODE REVIEW

**Comment by aaronjensen** [comment]
Nice, looking good so far.

**Comment by aaronjensen** [comment]
Might want to extract the hooks stuff to a common script that can be source by both `mergeq` and `mergeq_ci`, we'll want hooks in each. Hooks on the ci side are at least:

```
before_ci_commit
after_ci_push
```

I think `delete_feature_branch` can become an example hook, we probably shouldn't do it by default.

File: bin/mergeq.sh [hunk 1]
```diff
@@ -33,12 +80,25 @@ function exit_if_local_mods {
...
+}
+
 function merge_failed {
-  status "Doh. Your merge has conflicts, but don't worry:"
+  echo -e "${yellow}Doh. Your merge has conflicts, but don't worry:${default}"
   echo
   echo 1. Fix your merge conflicts
   echo 2. Commit them
-  echo 3. Run mergeq --continue
+  echo -e "3. Run ${blue}mergeq --continue${default}"
```
Comment (aaronjensen) [inline bin/mergeq.sh]: replace `mergeq` w/ `$0`

File: bin/mergeq.sh [hunk 2]
```diff
@@ -141,13 +213,13 @@ if [ "$target_branch" = "--continue" ] ; then
...
     continue_merge
   else
-    echo "
-**********************************************************
+    echo -e "
+${yellow}**********************************************************${default}
 
  It doesn't look like you're in the middle of a merge.
- Try mergeq <branch name> to start one
+ Try ${blue}mergeq <branch name>${default} to start one
```
Comment (aaronjensen) [inline bin/mergeq.sh]: same, `$0`

Comment (aaronjensen) [inline bin/mergeq.sh]: looks like there are a couple others.

File: bin/mergeq_ci.sh [hunk 3]
```diff
@@ -77,6 +101,8 @@ target_branch=$2
 
 validate_parameters
 
+run_hook "before_ci_startup" # check_dependencies
```
Comment (aaronjensen) [inline bin/mergeq_ci.sh]: I'm not sure if this one is even necessary, the dependencies are needed for one of the other hooks, so they could happen then. I suppose it's good to have any errors ahead of time though.

**Comment by dbalatero** [comment]
@aaronjensen I've queued this feature branch into `merge/staging` with `mergeq`. Now I want to test `mergeq_ci` locally as well on my laptop.

What branch state should I be in?

* on the exact sha1 of this "build" for `merge/staging`?
* on `merge/staging` itself?

I think it's the first one.

Also what should I pass into `mergeq_ci` as `$2`? 

* `staging`
* `merge/staging`

**Comment by aaronjensen** [comment]
```bash
git reset --hard <SHA of the "queing merge" commit>
mergeq_ci staging
```

**Comment by aaronjensen** [comment]
the "queing merge" commit is the tip of `merge/staging`

**Comment by dbalatero** [comment]
> the "queing merge" commit is the tip of merge/staging

Okay, I tested a full cycle of mergeq + merge_ci, and it's looking good. My branch made it to staging successfully.

**Comment by aaronjensen** [comment]
awesome!

File: bin/mergeq_common.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  hook_name=$1
+  hook="$hooks_dir/$hook_name"
+
+  [[ -f $hook ]] || return 0
+
+  status "Running hook: $hook_name..."
+
+  eval "$hook \"$target_branch\" \"$merge_branch\""
+  [[ $? -eq 0 ]] || exit $?
+}
```
Comment (aaronjensen) [inline bin/mergeq_common.sh]: I wonder if you should merge this back in and maybe surround it w/ comments so we know it is shared, just to keep the scripts self contained.

File: README.md [hunk 5]
```diff
@@ -16,9 +16,23 @@ Or install it yourself as:
...
+    $ echo ".mergeq/merging" >> .gitignore
+
+Create branches that you want to make queueable. This example creates a branch called
+`staging` for queuing builds.
+
+    $ cd your_project && git checkout master
+    $ git checkout -b staging
+    $ git push -u origin staging
+    $ git checkout -b merge/staging
+    $ git push -u origin merge/staging
```
Comment (aaronjensen) [inline README.md]: nit pick, but we don't necessarily need local branches:

```
$ git push origin master:staging
$ git push origin master:merge/staging
```

File: bin/mergeq_common.sh (hunk 4 above)
Comment (dbalatero) [inline bin/mergeq_common.sh]: Actually we should definitely, because the hooks will take different args…

Comment (aaronjensen) [inline bin/mergeq_common.sh]: ah, true. I'm working on porting this to tch right now so i can test it

File: README.md (hunk 5 above)
Comment (dbalatero) [inline README.md]: oh thanks, i was trying to remember how to do that

File: bin/mergeq.sh [hunk 6]
```diff
@@ -65,7 +128,7 @@ function try_to_merge {
 
 function write_temp_file {
   status "Writing temp file..."
-  echo "$branch;$merge_branch;$target_branch" > .merging
+  echo "$branch;$merge_branch;$target_branch" > $merging_file
```
Comment (aaronjensen) [inline bin/mergeq.sh]: need to mkdir -p the .mergeq folder here or the first run doesn't work

**Comment by aaronjensen** [comment]
successfully queued a merge on integration, build is running at the moment.

File: README.md [hunk 7]
```diff
@@ -16,9 +16,125 @@ Or install it yourself as:
...
+*`mergeq` hooks:*
+
+* `before_merge` - this runs before a merge happens
+* `after_push` - this runs after your branch is pushed to `merge/$target`
+* `after_cleanup` - this is the last thing that runs before `mergeq` exits
+
+*`mergeq_ci` hooks:*
+
+* `after_ci_merge` - this runs after CI merges into the target branch
+* `after_ci_push` - this runs after CI pushes the target branch back to origin
```
Comment (aaronjensen) [inline README.md]: before_ci_startup

**Comment by dbalatero** [comment]
@aaronjensen emailing you your new hooks

File: bin/mergeq_ci.sh [hunk 8]
```diff
@@ -65,6 +104,8 @@ function merge {
 function push {
   status "Pushing to $target_branch..."
   git push origin HEAD:$target_branch
+
+  run_hook "after_ci_push" # delete_feature_branch
```
Comment (aaronjensen) [inline bin/mergeq_ci.sh]: can get rid of these comments now

**Comment by aaronjensen** [comment]
build passed!

File: bin/mergeq.sh (hunk 6 above)
Comment (dbalatero) [inline bin/mergeq.sh]: :green_heart: d3678b3

**Comment by dbalatero** [comment]
@aaronjensen this is :package: now, pending more comments!

File: README.md [hunk 9]
```diff
@@ -16,9 +16,132 @@ Or install it yourself as:
...
-## Usage
+Setup your mergeq directory
 
-TODO: Write usage instructions here
+    $ cd your_project
+    $ mkdir .mergeq
+
+Ignore the merging lock file in git
+
+    $ echo ".mergeq/merging" >> .gitignore
```
Comment (dbalatero) [inline README.md]: :information_desk_person: It would be nice to just factor this all out into an idempotent setup script

**Comment by aaronjensen** [comment]
:shipit:! nice work and thanks @dbalatero

**Pull request merged**